dipy/denoise/gibbs.py

```diff
@@ -0,0 +1,314 @@
+
+    Note
+    ----
+    This function decreases the effects of gibbs oscilations based on the
+    analysis of local total variation (TV) along the two axis of the image.
+    Although artefact correction is done based on each point primary adjanced
+    neighbors, total variation should be accessed in a larger range of
+    neigbors. If you want to adjust the number of the neigbors to be
+    considered in TV calculation please change parameter nn.
+
+    References
+    ----------
+    Please cite the following articles
+    .. [1] Neto Henriques, R., 2018. Advanced Methods for Diffusion MRI Data
+           Analysis and their Application to the Healthy Ageing Brain
+           (Doctoral thesis). https://doi.org/10.17863/CAM.29356
+    .. [2] Kellner E, Dhital B, Kiselev VG, Reisert M. Gibbs-ringing artifact
+           removal based on local subvoxel-shifts. Magn Reson Med. 2016
+           doi: 10.1002/mrm.26054.
+    """
+    if np.any(G0) is None or np.any(G1) is None:
+        G0, G1 = _weights(image.shape)
+
+    img_c1 = _gibbs_removal_1d(image, axis=1, n_points=n_points)
+    img_c0 = _gibbs_removal_1d(image, axis=0, n_points=n_points)
+
+    C1 = np.fft.fft2(img_c1)
+    C0 = np.fft.fft2(img_c0)
+    imagec = abs(np.fft.ifft2(np.fft.fftshift(C1)*G1 + np.fft.fftshift(C0)*G0))
+
+    return imagec
+
+
+def gibbs_removal(vol, slice_axis=2, n_points=3):
+    """ Suppresses gibbs ringing artefacts of images volumes.
+
+    Parameters
+    ----------
+    vol : array ([X, Y, Z]) or ([X, Y, Z, g])
+        Matrix containing one volume (3D) or multiple (4D) volumes of images.
+    slice_axis : int (0, 1, or 2)
+        Data axis corresponding to the number of acquired slices. Default is
+        set to the third axis
+    n_points : int, optional
+        Number of neighbour points to access local TV (see note). Default is
+        set to 3.
+
+    Returns
+    -------
+    vol : array ([X, Y, Z]) or ([X, Y, Z, g])
+        Matrix containing one volume (3D) or multiple (4D) volumes of corrected
+        images.
+
+    Notes
+    -----
+    For 4D matrix last element should always correspond to the number of
+    diffusion gradient directions.
+
+    References
+    ----------
+    Please cite the following articles
+    .. [1] Neto Henriques, R., 2018. Advanced Methods for Diffusion MRI Data
+           Analysis and their Application to the Healthy Ageing Brain
+           (Doctoral thesis). https://doi.org/10.17863/CAM.29356
+    .. [2] Kellner E, Dhital B, Kiselev VG, Reisert M. Gibbs-ringing artifact
+           removal based on local subvoxel-shifts. Magn Reson Med. 2016
+           doi: 10.1002/mrm.26054.
+    """
+
+    # Make sure that slice are ordering in matrix axis 2
+    if slice_axis > 2:
+        raise ValueError("Different slices have to be organized along" +
+                         "one of the 3 first matrix dimensions")
+    elif slice_axis < 2:
+        vol = np.swapaxes(vol, slice_axis, 2)
```

Comment:
Data is already in the format assumed by the function when vol.ndim == 2, therefore you don't need that case in the elif statements. In the tests, I am testing a 2D image and all seems to run fine. I will add in the function documentation that data can also be a ([X, Y]) matrix.

Reply:
In your test, you do not swap axis with the 2d data and assume that the user will do the correct move :smile: 

if you do the same test with `swap_axis=0` (which is possible), this should crash. Since you are managing 2D data as well, it will be good to manage this case.

Reply:
Very well spotted! I've just fixed the issue! For 2D data, swap_axis is a dummy variable, because for 2D data you need just to run _gibbs_removal_2d. See more information in the comments of my code.

Reply:
Is this resolved? If yes, please hit the "resolve conversation" button